### ISSUE TITLE
Question: Reasonable upper limit on number of ENUM values?

### ISSUE THREAD

**Issue by slimdave** [rmosolgo/graphql-ruby#3498]
I have a situation where we deal with industry standard code lists that contain thousands of unique alphanumeric codes with descriptions. For example, [this one](https://ns.editeur.org/thema/en) has over 8,000.

We'd like to be able to use Enums, particularly when accepting these codes as arguments on a field or in an input type used in a mutation, as they automatically come with validation that way, and are available through introspection.

Basically, because it would very much reduce the amount of code we have to write and support in future.

Is it possible to offer any guidance on whether this is appropriate or not, and any circumstances that might affect that?

(Thanks very much for this gem - it is solving a lot of issues for us, and we very much appreciate the work that has gone into it).

**Comment by rmosolgo** [comment]
Glad to hear the library has been working ok so far! 

The only thing I'd say is that there's some code that doesn't scale well for huge enums, but it _could_. For example, this code enumerates through the enums one-by-one: 

https://github.com/rmosolgo/graphql-ruby/blob/9ecc20efadca8d499fe85e56748ab002f44b774b/lib/graphql/schema/enum.rb#L86-L87

But it could be refactored to maintain a hash of `graphql_name => value` instead, so that the lookup could use a hash table. 

Further down has the same issue: 

https://github.com/rmosolgo/graphql-ruby/blob/9ecc20efadca8d499fe85e56748ab002f44b774b/lib/graphql/schema/enum.rb#L99

But could use a similar fix to support huge enums.

This one _might_ need work, too: 

https://github.com/rmosolgo/graphql-ruby/blob/9ecc20efadca8d499fe85e56748ab002f44b774b/lib/graphql/schema/warden.rb#L136-L140

It creates a list of enum values which are allowed for _this_ query. It's only ever executed _once_, but it might be faster to use an approach that doesn't begin by iterating the whole list. 

In practice, it would probably only become a problem if: 

- You're accepting lots of enum inputs in one query. (Each input value would cause `.find { ... }` to be executed once -- so if you're accepting hundreds or thousands of enum values in a single query, it might add up.) 
- You're returning lots of enum values in one query. (Similarly, the second `.find { ... }` would be executed for each field that returns an enum value to a client.) 

Anyhow, my recommendation would be to write the schema the way you think it should be, then write a script to execute a could-be-problematic query with profiling (eg, stackprof) and share the profiling result here. With a real profiling result, we could be sure that I haven't overlooked anything (and besides, I might be _wrong_ about where the bottlenecks are!).

**Comment by slimdave** [comment]
Hmmm, very interesting.

I think that's a good recommendation - I'll see how performance looks with a large Enum. I think that it's unlikely that I'd want to put a lot of enum inputs into a single query ... maybe in the order of 10-20.

I'll look into stackprof.

Thanks very much Robert.

**Comment by rmosolgo** [comment]
Sure. I was thinking over this again last night and I'll leave some notes here in case it turns out there's work to do: 

One way this could be improved is to add some methods like `Enum#value?(graphql_name, context)` or `Enum#value_for(graphql_name, context)`, where, rather than returning a _list_, it returns single items. An API like that would support efficient implementations that don't require iterating and it would lay the groundwork for totally dynamic enums.

**Comment by slimdave** [comment]
So does that flip things round to something more like "is this item visible?" instead of "here's a list of all visible items, is this item in it", something like that?

**Comment by rmosolgo** [comment]
Sorry for the late response to your question, but yes, that's basically the idea. That way, we could support cases where "is this item visible?" can be performed faster than "here are all visible items?" (which, currently, involves generating a list of items, then checking the visibility of each one). 

If you run into trouble with big enums, please open a new issue to let me know, and we can dig in!
